### ISSUE TITLE
Validators should not make decisions based on which ports data was received

### ISSUE THREAD

**Issue by sagar-solana** [solana-labs/solana#6672]
#### Problem

Validators make many shred decisions based on which port it was received on. 
This is not secure at all. This issue is related to the regular DoS that is possible on any of a validator's ports. However it leaves the window open for an amplification attack. 

#### Proposed Solution

- For turbine Shreds, validators should do a reverse lookup to see if the Shred _should_ have arrived from where it did (i.e paired node in the layer above). If it did not match up then it shouldn't be sent through Turbine 
- For Repair Shreds, validators should only retransmit them if they came from the layer above. This can be coupled with the regular turbine shred described above. 
- For Repairman, validators should never send these shreds through turbine. No way to detect this right now. Needs a solution. 
- For all shreds received on the window, they should only be retransmitted if they have not already been inserted/recv'd

@aeyakovenko

**Comment by aeyakovenko** [comment]
Why isn’t the rule simply:

* retransmit once and only once if shred is from layer above.

Regardless how it arrived?  Where the leader is the layer above per node per shred.  It shouldn't matter how it arrived.

**Comment by sagar-solana** [comment]
Yeah that applies to everything except for Repairman. So Turbine doesn't care about where (in terms of slot) your peers are. Repairman is sending really old shreds (from the epoch before) to you and you should definitely not be sending them through turbine. Because your peers might not be as far behind as you are.

**Comment by sagar-solana** [comment]
So we need a solution for repairman. Something that isn't port dependent. 

Also We also send shreds received from neighbors to the children in layers below. So the rules should really be:
- Retransmit each shred just once. No matter where you received it from. This will involve a blocktree lookup. Layer above or below doesn't matter if you look at blocktree first. But this will have a performance impact (might need a cache)
- In addition to that, the scope of the retransmit will depend on whether the shred was destined to arrive from the layer above. If it was supposed to come from the layer above (doesn't matter if you repaired it) you should retransmit it neighbors and children. If it wasn't supposed to be delivered from the layer above, only send it to children.

**Comment by aeyakovenko** [comment]
@sagar-solana isn’t repairman request/response?

**Comment by sagar-solana** [comment]
@aeyakovenko no its not. It's automatic. Infact I just realized it repairs anyone more than 50 slots behind. Not an epoch anymore. So repairman looks at others' gossip and catches them up. No request. Directly response.

**Comment by aeyakovenko** [comment]
@sagar-solana but their request is signed by them.  so its a request to be repaired by the network which they can cancel.

**Comment by sagar-solana** [comment]
They can't take something off of gossip once it's put there? The response is not so much a response as much as it is just a flood of shreds on tvu. 

Anyway even if we allow nulling the gossipped EpochSlots that doesn't solve the main issue of not being able to differentiate shreds delivered via repairman vs turbine.
A validator never knows how far "behind" it is so it's not okay for it to treat both repairman and turbine shreds as equal. We need to be able to differentiate.

**Comment by aeyakovenko** [comment]
how does the flow control work with repairman?  does the node keep updating what it needs?

**Comment by sagar-solana** [comment]
@carllin please help clarify if I get this wrong.
@aeyakovenko, validators are always looking at everyone else's gossipped EpochSlots data. Which basically gets updated everytime a new root is made or new slots are completed.
The validators that are able to serve the slots to those nodes that can raise their root divvy up the work amongst each other and catchup the lagging validators.
No flow control really. Just every time a new epochslots shows up, they'll start dealing with it.

**Comment by mvines** [comment]
TODO: recover #7774 as a part of fixing this issue 😿

**Comment by behzadnouri** [comment]
Turbine port-based forwarding logic is removed as of https://github.com/solana-labs/solana/pull/17716 released on v1.7 branch
